### PR TITLE
Native weight resharding for Megatron RL

### PR DESCRIPTION
**Pull request by wdykas**
# What does this PR do ?
This is the initial starting for native fast weight resharding in Megatron. The fast path is being added it just might not be in this MR currently. 

Steps:
- [x] https://github.com/NVIDIA/Megatron-LM/pull/2485 for separations. Merge in changes
- [ ] NCCL & GLOO Baseline, This MR
- [ ] NVShmem Fast Path

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/transformer/utils.py [hunk 1]
```diff
@@ -29,13 +29,13 @@ def get_linear_layer(rows, columns, init_method, perform_initialization=True):
     return layer
 
 
-@lru_cache(maxsize=32)
+#@lru_cache(maxsize=32)
```
Comment (yobibyte) [inline megatron/core/transformer/utils.py]: Rollback? Or resharging won't work with lru_cache?

There's a hacky way to clear the cache which I use in rl_utils: https://github.com/tdene/Megatron-LM/blob/f3bb645b7d47f916a75b069f60b15211ceb9f096/megatron/rl/rl_utils.py#L2268

File: megatron/rl/inference/megatron.py [hunk 2]
```diff
@@ -100,6 +100,20 @@ def get_dynamic_inference_engine(args: Namespace, model: MegatronModule) -> Abst
 
     module = model.module.module if hasattr(model.module, "module") else model.module
 
+
+    # DynamicInferenceContext must use the inference model's TP size, not the
+    # training TP size from global args. The inference model may have a custom
+    # ProcessGroupCollection with a different TP size.
+    pg_collection = get_attr_wrapped_model(model, "pg_collection")
+    tp_group = getattr(pg_collection, 'tp', None) if pg_collection is not None else None
+    if tp_group is not None:
```
Comment (yobibyte) [inline megatron/rl/inference/megatron.py:122]: Trying to understand. In what situations, the tp_group is not going to have the pg_collection attribute?

File: megatron/rl/rl_utils.py [hunk 3]
```diff
@@ -2366,9 +2383,20 @@ def get_sequence_packing_tensorboard_metrics(args):
...
+        # TODO(Peter) We need to use the proper models MPU for refitting. If you forget you probably need to change this all over this 
+        # file
         bin_batch_size = (
             mpu.get_data_parallel_world_size() * args.micro_batch_size * get_num_microbatches()
         )
         metrics['bin-batch-size'] = bin_batch_size
         metrics['consumed-bins'] = args.consumed_train_bins
     return metrics
+
+def swap_train_to_inference_model(train_model: LanguageModule, inference_model: LanguageModule, refit_method: str):
```
Comment (yobibyte) [inline megatron/rl/rl_utils.py]: Do we need this function at all? Should we just call swap_model_weights directly?

File: megatron/training/arguments.py [hunk 4]
```diff
@@ -1972,6 +1972,13 @@ def _add_rl_args(parser):
                        help='Algorithm for distributing packed bins across ranks. '
                             'fifo: first-in-first-out sequential distribution, '
                             'round-robin: distribute bins cyclically across ranks for better load balancing')
+    group.add_argument('--rl-inference-tensor-model-parallel-size', type=int, default=None,
+                       help='Degree of tensor model parallelism for inference for RL.')     
+    group.add_argument('--refit-method', type=str, default='naive',
```
Comment (yobibyte) [inline megatron/training/arguments.py]: Make this --rl-refit-method to be consitent with the rest?

File: megatron/core/resharding/refit.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        else module.module
+        if hasattr(module, 'module')
+        else module
+    )
+
+
+def swap_model_weights(
+    src_model: LanguageModule,
+    target_model: LanguageModule,
+    refit_method: Union[str, CopyService],
```
Comment (yobibyte) [inline megatron/core/resharding/refit.py]: nit: can be Literal with predefined values to reduce the possible values.

File: megatron/core/resharding/copy_services/nccl_copy_service.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+import torch
+import torch.distributed as dist
+
+from .base import CopyService
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
```
Comment (yobibyte) [inline megatron/core/resharding/copy_services/nccl_copy_service.py:16]: nit: do these dataclasses belong to nccl_copy_service or better put in base?

File: megatron/training/training.py [hunk 7]
```diff
@@ -671,6 +674,46 @@ def pretrain(
     print_datetime('after model, optimizer, and learning rate ' 'scheduler are built')
     config = get_model_config(model[0])
 
+        # Build a separate inference model for RL if requested.
```
Comment (yobibyte) [inline megatron/training/training.py]: nit: Better wrap in a function to modularise training.py?

File: megatron/core/resharding/refit.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+from typing import Any, Optional, Union
+
+from megatron.core.models.common.language_module.language_module import LanguageModule
+from megatron.core import parallel_state
+from . import build_centralized_reshard_plan, execute_reshard_plan
+from .copy_services.base import CopyService
+from .copy_services.nccl_copy_service import NCCLCopyService
+
+
+def _unwrap_module(module: LanguageModule) -> Any:
```
Comment (yobibyte) [inline megatron/core/resharding/refit.py]: Would megatron.core.utils.unwrap_model work for you? 

https://github.com/NVIDIA/Megatron-LM/blob/ffb8c357e594f2eb14e71f6ce8831c0ffaf97f13/megatron/core/utils.py#L2055C5-L2055C17

File: megatron/core/resharding/utils.py [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+    try:
+        return group_ranks.index(global_rank)
+    except ValueError:
+        raise ValueError(
+            f"Rank {global_rank} not found in process group {group_ranks}. "
+            f"This likely indicates a configuration mismatch."
+        )
+
+
+def _detect_expert_index_from_param_name(param_name: str) -> Optional[int]:
```
Comment (yobibyte) [inline megatron/core/resharding/utils.py:110]: Add a unit test to future proof against naming changes?

File: megatron/core/resharding/utils.py [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+        key = tuple(m.data_parallel_group_ranks or [])
+        groups.setdefault(key, []).append(m)
+    if len(groups) == 1:
+        return src_meta_list[0]
+    dst_dp = dst_metadata.data_parallel_group_ranks or []
+    if dst_rank in dst_dp and len(dst_dp) > 0:
+        my_dst_dp_idx = dst_dp.index(dst_rank)
+    else:
+        my_dst_dp_idx = 0
+    keys_sorted = sorted(groups.keys())
```
Comment (yobibyte) [inline megatron/core/resharding/utils.py]: I'm not sure I understand these two lines, could you, please, explain why do we sort here?

File: megatron/core/resharding/planner.py [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+    dst_world = len(d.dst_dim_ranks)
+    src_local = src_shape[dim]
+    dst_local = dst_shape[dim]
+    if src_world * src_local != dst_world * dst_local:
+        raise RuntimeError(
+            f"{param_name}: size mismatch on TP dim{dim} "
+            f"(src_world={src_world}, src_local={src_local}, dst_world={dst_world}, dst_local={dst_local})"
+        )
+    # LCM tiling with strides
+    Ns = src_world * max(1, d.src_stride)
```
Comment (yobibyte) [inline megatron/core/resharding/planner.py:106]: nit: do we need max here? when we build this thing, we set it to 1 by default?

File: megatron/core/resharding/planner.py [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+            in_seg = g_micro % cps
+            src_owner_in_dim = s_idx % src_world
+            src_global_rank = d.src_dim_ranks[src_owner_in_dim]
+            src_local_seg_idx = s_idx // src_world
+            src_start = src_local_seg_idx * seg_src + in_seg * unit
+            dst_start = k * seg_dst + off * unit
+            # Build full N-D slices
+            src_slice = [slice(None)] * len(src_shape)
+            dst_slice = [slice(None)] * len(dst_shape)
+            src_slice[dim] = slice(src_start, src_start + unit)
```
Comment (yobibyte) [inline megatron/core/resharding/planner.py:139]: Can you explain the logic here? We fill in only one dim of the whole slice. Does these Nones get replaced later or are they used as is?

Should we unit-test this function separately?

File: megatron/core/resharding/execution.py [hunk 13]
```diff
@@ -0,0 +1,74 @@
+from __future__ import annotations
```
Comment (yobibyte) [inline megatron/core/resharding/execution.py:2]: CI will prob complain about lack of license in here and other files.

**Comment by wdykas** [comment]
/ok to test [9641c38](https://github.com/NVIDIA/Megatron-LM/pull/2379/commits/9641c38a6a2dde146109bf81ba33d03ede95383b)

**Comment by wdykas** [comment]
/ok to test [12dc7ae](https://github.com/NVIDIA/Megatron-LM/pull/2379/commits/12dc7ae19c7aa512a629046e3d3ae88055e5e5d0)

**Comment by wdykas** [comment]
/ok to test [70272da](https://github.com/NVIDIA/Megatron-LM/pull/2379/commits/70272da9719189d4af7dd9bd176b5ae0eca2e9d3)

File: megatron/training/arguments.py (hunk 4 above)
Comment (wdykas) [reply megatron/training/arguments.py]: I was leaving it like this because other people may want to use this outside RL

File: megatron/rl/inference/megatron.py (hunk 2 above)
Comment (wdykas) [reply megatron/rl/inference/megatron.py:122]: this is just a safeguard for backward compatibility. We may be able to remove

File: megatron/core/resharding/copy_services/nccl_copy_service.py (hunk 6 above)
Comment (wdykas) [reply megatron/core/resharding/copy_services/nccl_copy_service.py:16]: maybe the problem is we have more refit backends incoming in follow up MRs with different send signatures based on the copy service. I think we need to maybe unify it somehow. @kanz-nv do you have suggestions on how we can unify. The only worry here is the nvshmem backend requires extra meta data in its send/recieves

**Comment by wdykas** [comment]
/ok to test [cc5b44b](https://github.com/NVIDIA/Megatron-LM/pull/2379/commits/cc5b44bebf2a496ffe206a5a3f408396574a707f)

File: megatron/core/resharding/utils.py (hunk 10 above)
Comment (wdykas) [reply megatron/core/resharding/utils.py]: I simplified and commented the code a bit more

File: megatron/rl/rl_utils.py (hunk 3 above)
Comment (wdykas) [reply megatron/rl/rl_utils.py]: removed

File: megatron/core/resharding/refit.py (hunk 5 above)
Comment (wdykas) [reply megatron/core/resharding/refit.py]: added

File: megatron/core/transformer/cuda_graphs.py [hunk 14]
```diff
@@ -681,8 +715,10 @@ def create_fwd_graph(self, args, kwargs, clone_inputs=True):
         self.fwd_graph = torch.cuda.CUDAGraph()
 
         # For cases with multiple active RNG states, e.g. TP.
-        for _, state in get_all_rng_states().items():
-            self.fwd_graph.register_generator_state(state)
+        rng_states = get_all_rng_states()
```
Comment (wdykas) [inline megatron/core/transformer/cuda_graphs.py:718]: This change is necessary for when we want to do training cuda graphs. This should be going in at https://github.com/NVIDIA/Megatron-LM/pull/2452 so we can remove this once that goes in

File: megatron/rl/rl_utils.py [hunk 15]
```diff
@@ -1271,6 +1282,19 @@ def prepare_data_for_update(
...
     tb_writer = get_tensorboard_writer()
     nvtx_range = get_nvtx_range()
+
+    # RL policy updates + logprob computations should run eagerly; only rollout generation
+    # (inference engine) should use CUDA graphs until training cuda-graphs MR goes in. 
+    # In the single-model case this is naturally handled by `megatron_rl_inference_mode` 
+    # toggling graphs on/off around inference. In the refit case (separate inference_model),
+    # we must explicitly keep the training model (this `model`) with CUDA graphs disabled,
+    # otherwise training/logprobs can get cudagraphed.
+    if args.cuda_graph_impl != "none":
```
Comment (wdykas) [inline megatron/rl/rl_utils.py:1292]: This is for turning off training cuda graphs however https://github.com/NVIDIA/Megatron-LM/pull/2452 should support not having to do this so if that MR goes in first we should remove this
